File: 2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.py
class Solution:
    def countUnguarded(self, m: int, n: int, guards: List[List[int]], walls: List[List[int]]) -> int:
        arr = [[0 for _ in range(n)] for _ in range(m)]
        for r, c in walls:
            arr[r][c] = 1

        for r, c in guards:
            arr[r][c] = 2

        # for r, c in guards:
        #     rr = r
        #     while 0 < rr:
        #         rr -= 1
        #         if arr[rr][c] == 1 or arr[rr][c] == -1:
        #             break
        #         arr[rr][c] = -2
        #     rr = r
        #     while rr < m -1 :
        #         rr += 1
        #         if arr[rr][c] == 1 or arr[rr][c] == -1:
        #             break
        #         arr[rr][c] = -2
        #     cc = c
        #     while 0 < cc:
        #         cc -= 1
        #         if arr[r][cc] == 1 or arr[r][cc] == -1:
        #             break
        #         arr[r][cc] = -2
        #     cc = c
        #     while cc < n -1 :
        #         cc += 1
        #         if arr[r][cc] == 1 or arr[r][cc] == -1: 
        #             break
        #         arr[r][cc] = -2

        for i in range(m):
            guard = False
            for j in range(n):
                if arr[i][j] == 2:
                    guard = True
                elif arr[i][j] == 1:
                    guard = False
                if not arr[i][j] and guard:
                    arr[i][j] = 3

            guard = False
            for j in range(n - 1, -1, -1):
                if arr[i][j] == 2:
                    guard = True
                elif arr[i][j] == 1:
                    guard = False
                if not arr[i][j] and guard:
                    arr[i][j] = 3


        for j in range(n):
            guard = False
            for i in range(m):
                if arr[i][j] == 2:
                    guard = True
                elif arr[i][j] == 1:
                    guard = False
                if not arr[i][j] and guard:
                    arr[i][j] = 3

            guard = False
            for i in range(m - 1, -1, -1):
                if arr[i][j] == 2:
                    guard = True
                elif arr[i][j] == 1:
                    guard = False
                if not arr[i][j] and guard:
                    arr[i][j] = 3
        result = 0
        for i in arr:
            for j in i:
                if j == 0:
                    result += 1
        return result

File: 2343-count-unguarded-cells-in-the-grid/count-unguarded-cells-in-the-grid.java
class Solution {
    public void mark_guarded(int i, int j, int[][] arr, int m, int n){
        for (int t = i - 1; t >= 0; t--){
            if (arr[t][j] == 1 || arr[t][j] == 2)
                break;
            arr[t][j] = 3;
        }
        for (int t = i + 1; t < m; t++){
            if (arr[t][j] == 1 || arr[t][j] == 2)
                break;
            arr[t][j] = 3;
        }
        for (int t = j - 1; t >= 0; t--){
            if (arr[i][t] == 1 || arr[i][t] == 2)
                break;
            arr[i][t] = 3;
        }
        for (int t = j + 1; t < n; t++){
            if (arr[i][t] == 1 || arr[i][t] == 2)
                break;
            arr[i][t] = 3;
        }
    }

    public int countUnguarded(int m, int n, int[][] guards, int[][] walls) {

        int arr[][] = new int[m][n];
        for (int[] temp: arr)
            Arrays.fill(temp, 0);

        // 0 - Normal, 1 - Wall, 2 - Gaurd, 3 - Guard Sigth
        for (int i = 0; i < walls.length; i++){
            arr[walls[i][0]][walls[i][1]] = 1;
        }

        for (int i = 0; i < guards.length; i++){
            arr[guards[i][0]][guards[i][1]] = 2;
        }
        
        for (int i = 0; i < guards.length; i++){
            mark_guarded(guards[i][0], guards[i][1], arr, m, n);
        }


        int result = 0;

        for (int[] i: arr)
            for (int j: i)
                if (j == 0)
                    result += 1;

        return result;
    }
}